### PR TITLE
fix precompile array encode & other improvements

### PR DESCRIPTION
**Pull request by ermalkaleci**

### CODE REVIEW

**Comment by ermalkaleci** [comment]
> Is this a breaking change?

No. One of them is used on incentives precompiles which is still a PR and the other one is used by stableAsset precompile which doesn't have a contract yet

**Review by zjb0807 — DISMISSED** [review]

**Review by ferrell-code — APPROVED** [review]

File: runtime/common/src/precompile/input.rs [hunk 1]
```diff
@@ -201,68 +201,68 @@ where
 	}
 
 	fn bool_at(&self, index: usize) -> Result<bool, Self::Error> {
+		const ONE: U256 = U256([1u64, 0, 0, 0]);
 		let param = self.u256_at(index)?;
-		Ok(!param.is_zero())
+		if param == ONE {
```
[the remark this answers is not in the document]
Comment (ermalkaleci) [reply runtime/common/src/precompile/input.rs:206]: @ferrell-code actually I am trying an online decoder and it seems to handle non-zero as true. not sure now. will investigate more

[the remark this answers is not in the document]
Comment (ermalkaleci) [reply runtime/common/src/precompile/input.rs:206]: @ferrell-code ok tested. It needs to be zero or one.

[the remark this answers is not in the document]
Comment (ferrell-code) [reply runtime/common/src/precompile/input.rs:206]: :)

[the remark this answers is not in the document]
Comment (zjb0807) [reply runtime/common/src/precompile/input.rs:206]: https://github.com/rust-ethereum/ethabi/blob/master/ethabi/src/decoder.rs#L34-L40

We can use `ethabi::decode` to improve getting the input data.

**Review by zjb0807 — APPROVED** [review]

**Review by xlc — APPROVED** [review]

File: runtime/common/src/precompile/input.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ermalkaleci) [reply runtime/common/src/precompile/input.rs:206]: Yes I have thought about it

**Pull request merged**